### ISSUE TITLE
Occasionally, m2e is unable to read wrapped artifacts

### ISSUE THREAD

**Issue by pzi-dsa** [eclipse-m2e/m2e-core#743]
In our project, we make extensive use of the Spring framework which we add to our target platform via the m2e integration. And since they are only plain Maven artifacts, we let m2e rebundle them as OSGi plugins.
In general, this works quite well. But I frequently run into a situation, where Eclipse is unable to resolve those dependencies with an error similar to:

```
org.eclipse.core.runtime.CoreException: Error reading manifest for C:\Users\pzi\Desktop\eclipse-SDK-4.24M3-win32-x86_64\eclipse\configuration\org.eclipse.osgi\423\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\org.springframework\spring-context-support\5.3.20\spring-context-support-5.3.20.jar
	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:142)
	at org.eclipse.pde.core.target.TargetBundle.initialize(TargetBundle.java:171)
	at org.eclipse.pde.core.target.TargetBundle.<init>(TargetBundle.java:88)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.getWrappedArtifact(MavenTargetBundle.java:148)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.lambda$0(MavenTargetBundle.java:101)
	at org.eclipse.m2e.pde.target.CacheManager.accessArtifactFile(CacheManager.java:93)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.<init>(MavenTargetBundle.java:100)
	at org.eclipse.m2e.pde.target.CacheManager.getTargetBundle(CacheManager.java:117)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.addBundleForArtifact(MavenTargetLocation.java:372)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveDependency(MavenTargetLocation.java:278)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveArtifacts(MavenTargetLocation.java:177)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveBundles(MavenTargetLocation.java:156)
	at org.eclipse.pde.internal.core.target.AbstractBundleContainer.resolve(AbstractBundleContainer.java:96)
	at org.eclipse.pde.internal.core.target.TargetDefinition.resolve(TargetDefinition.java:424)
	at org.eclipse.pde.internal.ui.editor.targetdefinition.TargetEditor$TargetChangedListener$1.run(TargetEditor.java:606)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.io.EOFException: Unexpected end of ZLIB input stream
	at java.base/java.util.zip.ZipFile$ZipFileInflaterInputStream.fill(ZipFile.java:467)
	at java.base/java.util.zip.InflaterInputStream.read(InflaterInputStream.java:159)
	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:252)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:271)
	at org.eclipse.osgi.util.ManifestElement.readLine(ManifestElement.java:572)
	at org.eclipse.osgi.util.ManifestElement.parseBundleManifest(ManifestElement.java:512)
	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:113)
	... 15 more
```

Reloading the target platform has little effect. Sometimes it is able to load e.g. spring-context-support successfully, but then fails to load spring-context. The only way to fix it, is to restart the Eclipse.
Given that this is such an inconsistent behaviour, it is difficult to provide an example project.

Looking at the implementation, I noticed that rebundling is done by merging a modified Manifest into the original jar. Given the stack trace, my best assumption is now that this process somehow corrupts the final jar.

Do you have any other ideas what might cause this?

**Comment by laeubi** [comment]
Are you on linux or windows? In general I think one could catch the I/O error and create the wrapped item from scratch.

**Comment by laeubi** [comment]
> Windows 10.

it *might* be an issue that the jar-file is currently open and this the process can't overwrite it, so maybe it would be required to create a new file, try to delete the old and then rename it, if either delete an/or rename do not work just keep the new one...

**Comment by pzi-dsa** [comment]
I've updated my m2e installation after #765 was merged and I just got the following exception in the error log.

Good news: The additional logging is working as expected
Bad news: The problem still exists

```
eclipse.buildId=4.24.0.I20220601-1800
java.version=17.0.2
java.vendor=Oracle Corporation
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en_US
Command-line arguments:  -os win32 -ws win32 -arch x86_64

org.eclipse.m2e.pde.target
Error
Fri Jun 10 08:18:42 CEST 2022
net.sf.trove4j:trove4j:jar:2.1.0 is not a bundle and cannot be automatically bundled as such  (Error reading manifest for C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\419\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\net.sf.trove4j\trove4j\2.1.0\trove4j-2.1.0.jar)

org.eclipse.core.runtime.CoreException: Error reading manifest for C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\419\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\net.sf.trove4j\trove4j\2.1.0\trove4j-2.1.0.jar
	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:142)
	at org.eclipse.pde.core.target.TargetBundle.initialize(TargetBundle.java:171)
	at org.eclipse.pde.core.target.TargetBundle.<init>(TargetBundle.java:88)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.getWrappedArtifact(MavenTargetBundle.java:148)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.lambda$0(MavenTargetBundle.java:100)
	at org.eclipse.m2e.pde.target.CacheManager.accessArtifactFile(CacheManager.java:93)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.<init>(MavenTargetBundle.java:99)
	at org.eclipse.m2e.pde.target.CacheManager.getTargetBundle(CacheManager.java:117)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.addBundleForArtifact(MavenTargetLocation.java:372)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveDependency(MavenTargetLocation.java:278)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveArtifacts(MavenTargetLocation.java:177)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveBundles(MavenTargetLocation.java:156)
	at org.eclipse.pde.internal.core.target.AbstractBundleContainer.resolve(AbstractBundleContainer.java:96)
	at org.eclipse.pde.internal.core.target.TargetDefinition.resolve(TargetDefinition.java:428)
	at org.eclipse.pde.internal.ui.editor.targetdefinition.TargetEditor$TargetChangedListener$1.run(TargetEditor.java:606)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.io.EOFException: Unexpected end of ZLIB input stream
	at java.base/java.util.zip.ZipFile$ZipFileInflaterInputStream.fill(ZipFile.java:446)
	at java.base/java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:244)
	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:263)
	at org.eclipse.osgi.util.ManifestElement.readLine(ManifestElement.java:572)
	at org.eclipse.osgi.util.ManifestElement.parseBundleManifest(ManifestElement.java:512)
	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:113)
	... 15 more
```

So effectively, the same error as before.

I tried my hardest to reproduce the issue yesterday, but without success. But today, it showed up within minutes.
The only difference is that today, I'm connected to our company network via VPN, but given that the artifact already existed in my local Maven repository, I doubt that this is some kind of network issue where the artifact got corrupted while downloading.

**Comment by laeubi** [comment]
> I've updated my m2e installation after #765 was merged and I just got the following exception in the error log.
> 
> Good news: The additional logging is working as expected Bad news: The problem still exists
> 
> ```
> eclipse.buildId=4.24.0.I20220601-1800
> java.version=17.0.2
> java.vendor=Oracle Corporation
> BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en_US
> Command-line arguments:  -os win32 -ws win32 -arch x86_64
> 
> org.eclipse.m2e.pde.target
> Error
> Fri Jun 10 08:18:42 CEST 2022
> net.sf.trove4j:trove4j:jar:2.1.0 is not a bundle and cannot be automatically bundled as such  (Error reading manifest for C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\419\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\net.sf.trove4j\trove4j\2.1.0\trove4j-2.1.0.jar)
> 
> org.eclipse.core.runtime.CoreException: Error reading manifest for C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\419\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\net.sf.trove4j\trove4j\2.1.0\trove4j-2.1.0.jar
> 	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:142)
> 	at org.eclipse.pde.core.target.TargetBundle.initialize(TargetBundle.java:171)
> 	at org.eclipse.pde.core.target.TargetBundle.<init>(TargetBundle.java:88)
> 	at org.eclipse.m2e.pde.target.MavenTargetBundle.getWrappedArtifact(MavenTargetBundle.java:148)
> 	at org.eclipse.m2e.pde.target.MavenTargetBundle.lambda$0(MavenTargetBundle.java:100)
> 	at org.eclipse.m2e.pde.target.CacheManager.accessArtifactFile(CacheManager.java:93)
> 	at org.eclipse.m2e.pde.target.MavenTargetBundle.<init>(MavenTargetBundle.java:99)
> 	at org.eclipse.m2e.pde.target.CacheManager.getTargetBundle(CacheManager.java:117)
> 	at org.eclipse.m2e.pde.target.MavenTargetLocation.addBundleForArtifact(MavenTargetLocation.java:372)
> 	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveDependency(MavenTargetLocation.java:278)
> 	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveArtifacts(MavenTargetLocation.java:177)
> 	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveBundles(MavenTargetLocation.java:156)
> 	at org.eclipse.pde.internal.core.target.AbstractBundleContainer.resolve(AbstractBundleContainer.java:96)
> 	at org.eclipse.pde.internal.core.target.TargetDefinition.resolve(TargetDefinition.java:428)
> 	at org.eclipse.pde.internal.ui.editor.targetdefinition.TargetEditor$TargetChangedListener$1.run(TargetEditor.java:606)
> 	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
> Caused by: java.io.EOFException: Unexpected end of ZLIB input stream
> 	at java.base/java.util.zip.ZipFile$ZipFileInflaterInputStream.fill(ZipFile.java:446)
> 	at java.base/java.util.zip.InflaterInputStream.read(InflaterInputStream.java:158)
> 	at java.base/java.io.BufferedInputStream.fill(BufferedInputStream.java:244)
> 	at java.base/java.io.BufferedInputStream.read(BufferedInputStream.java:263)
> 	at org.eclipse.osgi.util.ManifestElement.readLine(ManifestElement.java:572)
> 	at org.eclipse.osgi.util.ManifestElement.parseBundleManifest(ManifestElement.java:512)
> 	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:113)
> 	... 15 more
> ```
> 
> So effectively, the same error as before.
> 
> I tried my hardest to reproduce the issue yesterday, but without success. But today, it showed up within minutes. The only difference is that today, I'm connected to our company network via VPN, but given that the artifact already existed in my local Maven repository, I doubt that this is some kind of network issue where the artifact got corrupted while downloading.

Any chance you can copy the offending jar and either upload or inspect its contents so we see whats wrong here?

**Comment by laeubi** [comment]
Funny enough 

```
<dependency>
    <groupId>net.sf.trove4j</groupId>
    <artifactId>trove4j</artifactId>
    <version>2.1.0</version>
</dependency>
```

is packed as `bundle` but without OSGi manifest headers :-1:

**Comment by laeubi** [comment]
> Good news: The additional logging is working as expected Bad news: The problem still exists

The Stacktrace is now even with m2e master code, that's also good :+1:

**Comment by pzi-dsa** [comment]
That's the jar in question. But after a quick inspection, I didn't see anything out of the ordinary.

[trove4j-2.1.0.jar.zip](https://github.com/eclipse-m2e/m2e-core/files/8876424/trove4j-2.1.0.jar.zip)

On further notice, I noticed a similar but unrelated exception in my error log, after restarting. Which is completely unrelated to m2e.

Perhaps this is a Java bug, rather than m2e/Eclipse? Because a quick search pointed me to this bug:
https://bugs.openjdk.org/browse/JDK-8283429
And given that I'm using Java 17.0.2, I'd still be affected by it.

```
eclipse.buildId=4.24.0.I20220601-1800
java.version=17.0.2
java.vendor=Oracle Corporation
BootLoader constants: OS=win32, ARCH=x86_64, WS=win32, NL=en_US
Command-line arguments:  -os win32 -ws win32 -arch x86_64

org.eclipse.pde.core
Error
Fri Jun 10 08:37:43 CEST 2022
Error reading manifest for C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\419\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\org.apache.lucene\lucene-core\9.2.0\lucene-core-9.2.0.jar

org.eclipse.core.runtime.CoreException: Error reading manifest for C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\419\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\org.apache.lucene\lucene-core\9.2.0\lucene-core-9.2.0.jar
	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:142)
	at org.eclipse.pde.internal.core.MinimalState.addBundle(MinimalState.java:194)
	at org.eclipse.pde.internal.core.PDEState.createNewTargetState(PDEState.java:166)
	at org.eclipse.pde.internal.core.PDEState.<init>(PDEState.java:61)
	at org.eclipse.pde.internal.core.PluginModelManager.initializeTable(PluginModelManager.java:616)
	at org.eclipse.pde.internal.core.PluginModelManager.targetReloaded(PluginModelManager.java:541)
	at org.eclipse.pde.internal.core.RequiredPluginsInitializer.lambda$0(RequiredPluginsInitializer.java:33)
	at org.eclipse.core.runtime.jobs.Job$2.run(Job.java:187)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.util.zip.ZipException: zip END header not found
	at java.base/java.util.zip.ZipFile$Source.findEND(ZipFile.java:1469)
	at java.base/java.util.zip.ZipFile$Source.initCEN(ZipFile.java:1477)
	at java.base/java.util.zip.ZipFile$Source.<init>(ZipFile.java:1315)
	at java.base/java.util.zip.ZipFile$Source.get(ZipFile.java:1277)
	at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:709)
	at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:243)
	at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:172)
	at org.eclipse.pde.internal.core.util.ManifestUtils.loadManifest(ManifestUtils.java:107)
	... 8 more
```

**Comment by laeubi** [comment]
> Perhaps this is a Java bug, rather than m2e/Eclipse? Because a quick search pointed me to this bug:
> https://bugs.openjdk.org/browse/JDK-8283429
> And given that I'm using Java 17.0.2, I'd still be affected by it.

That's really interesting! Can you try a java including the fix?

**Comment by pzi-dsa** [comment]
> Can you try a java including the fix?

It's seems like Java 17.0.4 won't come out before July 19th. I'm not sure if I get to build it locally, today, given that I'm not available for the next 2 weeks. I'll get back to you, once I've tested it, though.

One a side note, does Tycho wrap its artifacts in a similar way? Because this has never been an issue in our Jenkins builds.

**Comment by laeubi** [comment]
Tycho currently never caches an artifact but do it similar to m2e of course.

**Comment by pzi-dsa** [comment]
I've installed one of the recent [nightly builds](https://adoptium.net/de/temurin/nightly/) of Java 17.0.4 and the problem still shows up.

Given that this primarily happens when I'm using my VPN connection, perhaps this really is just a network problem where the file gets corrupted during the download.

**Comment by laeubi** [comment]
I'm just wondering why a refresh do not work? Maybe we could/should enable maven to check the file signatures here (md5/sha)?

**Comment by pzi-dsa** [comment]
> Maybe we could/should enable maven to check the file signatures here (md5/sha)?

That was my idea, as well. Just to double-check whether this is actually the problem.

**Comment by pzi-dsa** [comment]
A short update: I compared sha-hash of the original artifacts with the hash in the .sha1 file. Both are identical, so the original jar is valid. I somewhat expected this since otherwise, the bndtools library would most likely thrown an error when patching the Manifest.

```
SHA-1 for spring-core: 4b88aa3c401ede3d6c8ac78ea0c646cf326ec24b, expected 4b88aa3c401ede3d6c8ac78ea0c646cf326ec24b
```

I then wrapped the call to `new MavenTargetBundle(...)` in a try-catch block, where I call FileUtils.forceDelete() whenever an exception is thrown. The idea being that corrupted artifacts should be deleted, so that a fresh file is created when clicking the reload button.
However, this doesn't work, as the file is locked by another process.

```
java.io.IOException: Unable to delete file: C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\437\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\org.springframework\spring-core\5.3.20\spring-core-5.3.20.jar
	at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1425)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.getWrappedArtifact(MavenTargetBundle.java:163)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.lambda$0(MavenTargetBundle.java:107)
	at org.eclipse.m2e.pde.target.CacheManager.accessArtifactFile(CacheManager.java:93)
	at org.eclipse.m2e.pde.target.MavenTargetBundle.<init>(MavenTargetBundle.java:106)
	at org.eclipse.m2e.pde.target.CacheManager.getTargetBundle(CacheManager.java:117)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.addBundleForArtifact(MavenTargetLocation.java:372)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveDependency(MavenTargetLocation.java:278)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveArtifacts(MavenTargetLocation.java:177)
	at org.eclipse.m2e.pde.target.MavenTargetLocation.resolveBundles(MavenTargetLocation.java:156)
	at org.eclipse.pde.internal.core.target.AbstractBundleContainer.resolve(AbstractBundleContainer.java:96)
	at org.eclipse.pde.internal.core.target.TargetDefinition.resolve(TargetDefinition.java:428)
	at org.eclipse.pde.internal.ui.editor.targetdefinition.TargetEditor$TargetChangedListener$1.run(TargetEditor.java:606)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.nio.file.FileSystemException: C:\Users\pzi\Desktop\eclipse-SDK-4.24RC2-win32-x86_64\eclipse\configuration\org.eclipse.osgi\437\data\29c371a4db03d06c9603cb81c10fea468c03e5d6\org.springframework\spring-core\5.3.20\spring-core-5.3.20.jar: The process cannot access the file because it is being used by another process
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:275)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:110)
	at java.base/java.nio.file.Files.deleteIfExists(Files.java:1191)
	at org.apache.commons.io.file.PathUtils.deleteFile(PathUtils.java:365)
	at org.apache.commons.io.file.PathUtils.delete(PathUtils.java:304)
	at org.apache.commons.io.file.PathUtils.delete(PathUtils.java:280)
	at org.apache.commons.io.FileUtils.forceDelete(FileUtils.java:1423)
	... 13 more
```

> t might be an issue that the jar-file is currently open and this the process can't overwrite it

It looks like that guess is correct and that there is a concurrent access to the file. Maybe I can figure out which processes exactly keep the lock. But otherwise, the easiest solution really seems to be to save the jar in a new file and then try to replace it with the original file, as you suggested.

**Comment by pzi-dsa** [comment]
> it might be an issue that the jar-file is currently open and this the process can't overwrite it, so maybe it would be required to create a new file, try to delete the old and then rename it, if either delete an/or rename do not work just keep the new one...

I had a look at it again, and seems like the "other process" is indeed the Eclipse (or rather the Target Editor, to be more specific).
I tested a patch over the past few weeks, where I created a new file, whenever the bundled jar needs to be refreshed and since then, the problem never appeared again.

**Comment by pzi-dsa** [comment]
Closed together with #756
